src/metrics/detect_covariance.nit

```diff
@@ -0,0 +1,506 @@
+	fun cpt_total_variance: Counter[String] do return once new Counter[String]
+	fun cpt_total_classes: Counter[String] do return once new Counter[String]
+
+	fun cpt_explanations: Counter[String] do return once new Counter[String]
+	fun cpt_classes: Counter[MClass] do return once new Counter[MClass]
+	fun cpt_pattern: Counter[String] do return once new Counter[String]
+	fun cpt_nodes: Counter[String] do return once new Counter[String]
+	fun cpt_modules: Counter[String] do return once new Counter[String]
+	fun cpt_expression: Counter[String] do return once new Counter[String]
+
+	fun cpt_cast_kind: Counter[String] do return once new Counter[String]
+	fun cpt_cast_classes: Counter[String] do return once new Counter[String]
+	fun cpt_cast_pattern: Counter[String] do return once new Counter[String]
+	fun cpt_autocast: Counter[String] do return once new Counter[String]
+
+	# Display collected statistics
+	redef fun process_mainmodule(mainmodule, given_mmodules)
+	do
+		if is_disabled then return
+
+		print "--- Detection of the usage of covariance static type conformance ---"
+
+		print "-- Total --"
+		print "- Kinds of the subtype -"
+		cpt_subtype_kinds.print_elements(10)
+		print "  total: {cpt_subtype_kinds.sum}"
+
+		print "- Variance -"
+		cpt_total_variance.print_elements(10)
+		print "  total: {cpt_total_variance.sum}"
+
+		print "- Classes of the subtype -"
+		cpt_total_classes.print_elements(10)
+		print "  total: {cpt_total_classes.sum}"
+
+		print "-- On covariance only --"
+
+		print "- Specific covariance case explanations -"
+		cpt_explanations.print_elements(10)
+		print "  total: {cpt_explanations.sum}"
+
+		print "- Classes of the subtype, when covariance -"
+		cpt_classes.print_elements(10)
+		print "  total: {cpt_classes.sum}"
+
+		print "- Patterns of the covariant cases -"
+		cpt_pattern.print_elements(10)
+		print "  total: {cpt_pattern.sum}"
+
+		print "- Nodes of the covariance cases -"
+		cpt_nodes.print_elements(10)
+		print "  total: {cpt_nodes.sum}"
+
+		print "- Modules of the covariance cases -"
+		cpt_modules.print_elements(10)
+		print "  total: {cpt_modules.sum}"
+
+		print "- Kind of the expression node (when it make sense) -"
+		cpt_expression.print_elements(10)
+		print "  total: {cpt_expression.sum}"
+
+		print "-- Casts --"
+
+		print "- Kind of cast target -"
+		cpt_cast_kind.print_elements(10)
+		print "  total: {cpt_cast_kind.sum}"
+
+		print "- Classes of the cast -"
+		cpt_cast_classes.print_elements(10)
+		print "  total: {cpt_cast_classes.sum}"
+
+		print "- Cast pattern -"
+		cpt_cast_pattern.print_elements(10)
+		print "  total: {cpt_cast_pattern.sum}"
+
+		print "- Autocasts -"
+		cpt_autocast.print_elements(10)
+		print "  total: {cpt_autocast.sum}"
+	end
+
+	# Common method used when static subtype test is performed by a phase
+	# Returns true if the test concern real generic covariance
+	fun count_types(node: ANode, elem: ANode, sub, sup: MType, mmodule: MModule, anchor: nullable MClassType): Bool
```

Comment:
You combined `sub` and `sup` types but not `node` and `elem`.

This kind of bothers me.